src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -2336,47 +2345,107 @@ void StorageReplicatedMergeTree::cloneReplica(const String & source_replica, Coo
+    size_t max_log_entry = 0;
+    if (!log_entries.empty())
+    {
+        String last_entry_num = std::max_element(log_entries.begin(), log_entries.end())->substr(4);
+        max_log_entry = std::stol(last_entry_num);
+    }
+    /// log_pointer can point to future entry, which was not created yet
+    ++max_log_entry;
 
+    size_t min_replication_lag = std::numeric_limits<size_t>::max();
     String source_replica;
     Coordination::Stat source_is_lost_stat;
-    source_is_lost_stat.version = -1;
-
-    for (const String & source_replica_name : zookeeper->getChildren(zookeeper_path + "/replicas"))
+    size_t future_num = 0;
+    for (const String & source_replica_name : replicas)
     {
-        String source_replica_path = zookeeper_path + "/replicas/" + source_replica_name;
+        if (source_replica_name == replica_name)
+            continue;
 
-        /// Do not clone from myself.
-        if (source_replica_path != replica_path)
+        auto get_is_lost     = futures[future_num++].get();
+        auto get_log_pointer = futures[future_num++].get();
+        auto get_queue       = futures[future_num++].get();
+
+        if (get_is_lost.error == Coordination::Error::ZNONODE)
         {
-            /// Do not clone from lost replicas.
-            String source_replica_is_lost_value;
-            if (!zookeeper->tryGet(source_replica_path + "/is_lost", source_replica_is_lost_value, &source_is_lost_stat)
-                || source_replica_is_lost_value == "0")
-            {
-                source_replica = source_replica_name;
-                break;
-            }
+            /// For compatibility with older ClickHouse versions
+            get_is_lost.stat.version = -1;
+        }
+        else if (get_is_lost.data != "0")
+            continue;
+        if (get_log_pointer.error != Coordination::Error::ZOK)
+            continue;
+        if (get_queue.error != Coordination::Error::ZOK)
+            continue;
+
+        /// Replica is not lost and we can clone it. Let's calculate approx replication lag.
+        size_t source_log_pointer = get_log_pointer.data.empty() ? 0 : parse<UInt64>(get_log_pointer.data);
+        assert(source_log_pointer <= max_log_entry);
+        size_t replica_queue_lag = max_log_entry - source_log_pointer;
+        size_t replica_queue_size = get_queue.stat.numChildren;
+        size_t replication_lag = replica_queue_lag + replica_queue_size;
+        LOG_INFO(log, "Replica {} has approximate {} queue lag and {} queue size", source_replica_name, replica_queue_lag, replica_queue_size);
+        if (replication_lag < min_replication_lag)
+        {
+            source_replica = source_replica_name;
+            source_is_lost_stat = get_is_lost.stat;
+            min_replication_lag = replication_lag;
         }
     }
 
     if (source_replica.empty())
         throw Exception("All replicas are lost", ErrorCodes::ALL_REPLICAS_LOST);
 
+    if (is_new_replica)
```

Comment:
let's also add info about source queue size and log pointer.

Reply:
Maybe add log pointer here?
https://github.com/ClickHouse/ClickHouse/blob/47e9fe9ff16aa48c3f9b77ea1e9764b9cc17af20/src/Storages/StorageReplicatedMergeTree.cpp#L2432